test/unit/org/apache/cassandra/index/sai/cql/VectorUpdateDeleteTest.java

```diff
@@ -780,6 +780,39 @@ public void ensureVariableChunkSizeDoesNotLeadToIncorrectResults() throws Except
         }
     }
 
+    @Test
+    public void shadowedPrimaryKeysRequireDeeperSearch() throws Throwable
+    {
+        createTable(KEYSPACE, "CREATE TABLE %s (pk int primary key, str_val text, val vector<float, 2>)");
+        createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex'");
+        createIndex("CREATE CUSTOM INDEX ON %s(str_val) USING 'StorageAttachedIndex'");
+        waitForIndexQueryable();
+        disableCompaction(KEYSPACE);
+
+        // Create 100 rows so that each row has a slightly less similar score.
+        for (int i = 0; i < 100; i++)
+            execute("INSERT INTO %s (pk, str_val, val) VALUES (?, 'A', ?)", i, vector(1, i));
+
+        flush();
+
+        // Create 10 rows with the worst scores, but they won't be shadowed.
+        for (int i = 100; i < 110; i++)
+            execute("INSERT INTO %s (pk, str_val, val) VALUES (?, 'A', [1,1000])", i);
+
+        // Delete first 90 rows
+        for (int i = 0; i < 90; i++)
+            execute("DELETE FROM %s WHERE pk = ?", i);
+
+        beforeAndAfterFlush(() -> {
+            // ANN Only
+            assertRows(execute("SELECT pk FROM %s ORDER BY val ann of [1.0, 1.0] LIMIT 4"),
+                       row(90), row(91), row(92), row(93));
```

Comment:
One note about the `softLimit` logic: this query appears to search 24 total times because we're not increasing the shadowed limit sufficiently.

Reply:
I thought Piotr addressed that.  is this a round-down instead of round-up problem?

Reply:
I'm not sure yet, looking.

Reply:
@pkolaczk - would you be able to take a look at this test and the `currentSoftLimitEstimate()` logic? It looks like the test's execution always chooses `target + shadowedCount`.

```java
        int limit = Math.max(target + shadowedCount, Math.min(uncappedLimit, prevSoftLimit * 10));
```

Our previous logic was `K + 2 * M` or `target + 2 * shadowedCount`. When I use that for the test, it completes in 5 passes.

However, I'm not sure that we want to optimize for this specific test case. It feels like a contrived example. What are your thoughts?

Reply:
The previous logic was either `target + shadowedCount` or `target + shadowedCount * 2` depending on the flag for speculative logic. It might be just a coincidence `shadowedCount * 2` works as it is simply higher. I guess what might be happening is we misestimate the probability of getting shadowed keys somehow? I'm taking a look now. Thank you for the test.

Reply:
After the fix it looks like this:
```
DEBUG [main] 2024-01-09 14:30:50,378 QueryController.java:585 - Soft limit estimate: 52 with target=4 shadowed=0/0 P=0.125
DEBUG [main] 2024-01-09 14:30:50,378 QueryController.java:424 - getTopKRows using limit = 52
DEBUG [main] 2024-01-09 14:30:50,384 StorageAttachedIndexSearcher.java:140 - Shadow loop iteration 1: rows matched: 0, keys shadowed: 52
DEBUG [main] 2024-01-09 14:30:50,385 QueryController.java:585 - Soft limit estimate: 693 with target=4 shadowed=52/52 P=0.009615385
DEBUG [main] 2024-01-09 14:30:50,385 QueryController.java:424 - getTopKRows using limit = 693
DEBUG [main] 2024-01-09 14:30:50,391 StorageAttachedIndexSearcher.java:140 - Shadow loop iteration 2: rows matched: 603, keys shadowed: 90
DEBUG [main] 2024-01-09 14:30:50,392 QueryController.java:307 - Decided filter sort order SORT_THEN_FILTER (costs: SORT_THEN_FILTER = 800.0, FILTER_THEN_SORT = 910.0)
DEBUG [main] 2024-01-09 14:30:50,392 QueryController.java:585 - Soft limit estimate: 4 with target=4 shadowed=0/0 P=1.0
DEBUG [main] 2024-01-09 14:30:50,393 QueryController.java:424 - getTopKRows using limit = 4
DEBUG [main] 2024-01-09 14:30:50,394 StorageAttachedIndexSearcher.java:140 - Shadow loop iteration 1: rows matched: 0, keys shadowed: 4
DEBUG [main] 2024-01-09 14:30:50,394 QueryController.java:585 - Soft limit estimate: 52 with target=4 shadowed=4/4 P=0.125
DEBUG [main] 2024-01-09 14:30:50,394 QueryController.java:424 - getTopKRows using limit = 52
DEBUG [main] 2024-01-09 14:30:50,397 StorageAttachedIndexSearcher.java:140 - Shadow loop iteration 2: rows matched: 0, keys shadowed: 52
DEBUG [main] 2024-01-09 14:30:50,397 QueryController.java:585 - Soft limit estimate: 693 with target=4 shadowed=52/52 P=0.009615385
DEBUG [main] 2024-01-09 14:30:50,397 QueryController.java:424 - getTopKRows using limit = 693
DEBUG [main] 2024-01-09 14:30:50,401 StorageAttachedIndexSearcher.java:140 - Shadow loop iteration 3: rows matched: 603, keys shadowed: 90
```